sda/protocol.go

```diff
@@ -100,3 +101,99 @@ func ProtocolNameToID(name string) ProtocolID {
 func GlobalProtocolRegister(name string, protocol NewProtocol) (ProtocolID, error) {
 	return protocols.Register(name, protocol)
 }
+
+// ProtocolIO is an interface that allows one protocol to completely define its
+// wire protocol format while still using the Overlay. Implementations must
+// provide methods to read a packet coming from the network and also to write a
+// packet going to the network. A default one is provided with
+// defaultProtocolIO so the regular wire-format protocol can still be used.
+type ProtocolIO interface {
+	// Wrap takes a message and the overlay information and returns the message
+	// has to be sent directly to the network alongside with any error that
+	// happened.
+	// the bigger outer struct. Msg can be nil, that case it means the message
+	// is only an internal message of the Overlay.
+	Wrap(msg interface{}, info *OverlayMessage) (interface{}, error)
+	// Unwrap takes the message coming from the network and must returns the
+	// inner message that is going to be dispatched to the ProtocolInstance, the
+	// OverlayMessage needed by the Overlay to function correctly and then any
+	// error that might have occured.
+	Unwrap(msg interface{}) (interface{}, *OverlayMessage, error)
+	// PacketType returns the packet type ID that this Protocol expects from the
+	// network. This is needed in order for the Overlay to receive those
+	// messages and dispatch them to the correct ProtocolIO.
+	PacketType() network.PacketTypeID
+}
+
+// NewProtocolIO is a function typedef to instantiate a new ProtocolIO
+type NewProtocolIO func() ProtocolIO
+
+type protocolIOFactoryStruct struct {
+	factories map[string]NewProtocolIO
+}
+
+var protocolIOFactory = protocolIOFactoryStruct{
+	factories: make(map[string]NewProtocolIO),
+}
+
+// RegisterProtocolIO takes a name and and NewProtocolIO and save both fields.
+// When a Conode is instantiated, all ProtocolIO will be generated and stored
+// for this Conode.
+func RegisterProtocolIO(name string, n NewProtocolIO) {
+	_, present := protocolIOFactory.factories[name]
+	if present {
+		log.Error("protocolIOStore already registered a ProtocolIO at this name", name)
+		return
+	}
+	protocolIOFactory.factories[name] = n
+}
+
+// protocolIOStore contains all created ProtocolIO and is generally used by the
+// Overlay. It contains the default ProtocolIO used by the Overlay in order to
+// still function properly in case the old wire-format protocol is used.
+type protocolIOStore struct {
+	sync.Mutex
+	protos []ProtocolIO
+	names  map[string]int
+	types  map[network.PacketTypeID]int
+	// the one that gets used in case no ProtocolIO is defined
+	defaultIO ProtocolIO
```

Comment:
Why is this not a `ProtocolIO` stored under `default`? With this solution we have two places where `ProtocolIO`s are stored.